### ISSUE TITLE
Human readable @context

### ISSUE THREAD

**Issue by David-Chadwick** [w3c/vc-data-model#585]
This is the text I am questioning:

A human-readable document describing the expected order of values for the
@context property is expected to be published by any
implementer seeking interoperability. A machine-readable description
(i.e., a normal JSON-LD Context document) is expected to be published
at the URL specified in the @context property by
JSON-LD implementers seeking interoperability.

I have one question about this text which is, how does a JSON processor find the human readable document that is referred to? The text only says that the machine-readable description should exist at the URL in the @context. Wouldn't it be better to say that the human readable version should also be at the same URL, perhaps as comments in the JSON-LD.

**Comment by msporny** [comment]
> I have one question about this text which is, how does a JSON processor find the human readable document that is referred to?

A JSON processor doesn't find the human-readable document, only a human can find that. They typically find it the way they find other documents for JSON processors today... which is by doing a Google Search.

> Wouldn't it be better to say that the human readable version should also be at the same URL, perhaps as comments in the JSON-LD.

JSON doesn't support comments. :)

One could Content Negotiate for a human-readable version, but we have historically hesitated to do that because developers often misimplement content negotiation leading to terrible interop stories and wailing in blog posts about how terrible conneg is. :)

> The text only says that the machine-readable description should exist at the URL in the @context.

It's a fairly common practice for Linked Data folks to copy the expanded terms, which are URLs, and paste them into a browser location bar, which brings up documents such as this:

https://www.w3.org/2018/credentials#issuer

Does that help explain? Based on that, what changes do you think need to be made to the specification?

**Comment by gkellogg** [comment]
The JSON-LD 1.1 spec will allow JSON-LD contexts to be extracted from script tags in HTML documents for precisely this purpose (see https://w3c.github.io/json-ld-syntax/#using-an-html-document-as-a-context). One such document was published at https://json-ld.org/contexts/person.html. As this is something that may not be necessary in certain embedded environments, the notion of processor classes was introduced to allow a **[pure JSON Processor](https://w3c.github.io/json-ld-syntax/#dfn-pure-json-processor)** to conform without processing HTML. But, a **[full Processor](https://w3c.github.io/json-ld-syntax/#dfn-full-processor)** is expected to do this.

For example, a context in an HTML document would be something like the following:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
    <title>Context definition of a person</title>
    <script type="application/ld+json;profile=http://www.w3.org/ns/json-ld#context">
      {
         "@context": {...}
       }
    </script>
  </head>
  <body>
    ...
  </body>
</html>
```

**Comment by David-Chadwick** [comment]
@gkellogg I like your solution and the example of the person @context that you provided at https://json-ld.org/contexts/person.html provides both the human readable information that is needed, as well as machine processable version. 
@manu could we add text to say something like "It is recommended that script tags in HTML are used so that both human readable and machine processable @contexts can be published at the same URL. For example.... <include here @gkellog's example> "

**Comment by msporny** [comment]
> As this is something that may not be necessary in certain embedded environments, the notion of processor classes was introduced to allow a pure JSON Processor to conform without processing HTML. But, a full Processor is expected to do this.

To be clear, I really dislike this feature of JSON-LD 1.1 because it raises the burden of Full JSON-LD 1.1 processors to contain an HTML processor (which is a massive requirement) on top of doing JSON-LD processing. I also think this is going to really damage the adoption of JSON-LD 1.1 and make it so much easier for people to argue against it... hell, even I would argue against "Full JSON-LD processors" (and plan to if this feature goes to REC).

Note that we WG had a hard enough time building bridges between JSON developers that didn't even want to pull in a JSON-LD processor... and now we're suggesting that they pull in a full blown HTML processor as well?

We have constrained device use cases... doing this would add complexity where the benefit is unclear.

**Comment by gkellogg** [comment]
> > As this is something that may not be necessary in certain embedded environments, the notion of processor classes was introduced to allow a pure JSON Processor to conform without processing HTML. But, a full Processor is expected to do this.
> 
> To be clear, I really dislike this feature of JSON-LD 1.1 because it raises the burden of Full JSON-LD 1.1 processors to contain an HTML processor (which is a massive requirement) on top of doing JSON-LD processing. I also think this is going to really damage the adoption of JSON-LD 1.1 and make it so much easier for people to argue against it... hell, even I would argue against "Full JSON-LD processors" (and plan to if this feature goes to REC).

I appreciate your position, but JSON-LD in HTML is probably the biggest use case right now (although that will likely change with adoption of VC and WoT). JSON-LD in HTML is a reality that the spec needs to recognize and legitimize.

Of course, a full HTML processor is not required, just a parser. These are widely available, and the problem with integrating them together is largely that of the size of the resulting implementation, rather than the complexity of extracting JSON-LD from script elements.

There is certainly room to finesse the processor levels, but having a mode for processors that wish to properly handle JSON-LD embedded in HTML is important.

Of course, the best place to raise these concerns is in the JSON-LD WG, but so far, the concerns have not outweighed the benefits for most of the WG.

**Comment by msporny** [comment]
I have requested the removal of the "JSON-LD Context in HTML document" feature in JSON-LD 1.1. That conversation is happening in that issue tracker: https://github.com/w3c/json-ld-syntax/issues/172

**Comment by msporny** [comment]
Add something to implementation guidance to define how one publishes a VC Context extension that is human readable.

**Comment by msporny** [comment]
Added section on human readability for extensions in commit https://github.com/w3c/vc-imp-guide/commit/504c9c1189d060cc9f1deb2ff35507e1fbf3f282.

Suggest that we close this issue as nothing else is likely going to happen to address this issue in the near future.

**Comment by David-Chadwick** [comment]
OK, the implementation guide seems a reasonable way of resolving this. So we can close this issue

**Comment by burnburn** [comment]
WG resolution on 28 May 2019 call: 

RESOLUTION: Issue #585 is resolved via commit 504c9c1189d060cc9f1deb2ff35507e1fbf3f282 to the Implementation Guide, which advises JSON-LD Context publishers on how to publish human readable contexts that advise on context order. Issue #585 should be closed after a 7 day review period.

**Comment by TallTed** [comment]
For clarity, [commit 504c9c1189d060cc9f1deb2ff35507e1fbf3f282 to the Implementation Guide](https://github.com/w3c/vc-imp-guide/commit/504c9c1189d060cc9f1deb2ff35507e1fbf3f282).
